working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,937 @@
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+A compile-time error occurs if an extension type is used as a
+superinterface of a class or mixin, or if an extension type is used to
+derive a mixin.
+
+*So `class C extends E1 with E2 implements E3 {}` has three errors if `E1`,
+`E2`, and `E3` are extension types, and `mixin M on E1 implements E2 {}`
+has two errors.*
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is treated as
+<code>invokeExtensionMethod(Ext<S<sub>1</sub>, .. S<sub>k</sub>>, e).m(args)</code>
+and similarly for instance getters and operators. This rule also applies
+when a member access implicitly has the receiver `this`, and the static
+type of `this` is an extension type (*which can only occur in an extension
+type member declaration where the on-type is itself an extension type*).
+
+*That is, when the type of an expression is an extension type, all method
+invocations on that expression will invoke an extension method declared by
+that extension, and similarly for other member accesses. In particular, we
+cannot invoke an instance member when the receiver type is an extension
+type (unless the the extension type enables them explicitly, cf. the
+show/hide part specified in a later section).*
+
+Let `E` be an extension declaration named `Ext` with type parameters
+<code>X<sub>1</sub> extends B<sub>1</sub>, .. X<sub>k</sub> extends B<sub>k</sub></code>
+and on-type clause `on T`. Then we say that the _declared on-type_ of `Ext`
+is `T`, and the _instantiated on-type_ of
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>
+is
+<code>[S<sub>1</sub>/X<sub>1</sub>, .. S<sub>k</sub>/X<sub>k</sub>]T</code>.
+We will omit 'declared' and 'instantiated' from the phrase when it is clear
+from the context whether we are talking about the extension itself or a
+particular instantiation of a generic extension. For non-generic
+extensions, the on-type is the same in either case.
+
+An extension type `E` of the form
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>
+is a subtype of `Object?`, and a proper supertype of the
+instantiated on-type of `E`. In the case where the instantiated on-type of
+`E` is not a top type, `E` is also a proper subtype of `Object?`.
+
+*That is, the underlying on-type can only be recovered by an explicit cast
+(except when the on-type is a top type). So an expression whose type is an
+extension type is in a sense "in prison", and we can only obtain a
+different type for it by forgetting everything (going to a top type), or by
+means of an explicit cast, typically a downcast to the on-type.*
+
+When `E` is an extension type, a type test `o is E` or `o is! E` and a type
+check `o as E` can be performed. Such checks performed on a local variable
+can promote the variable to the extension type using the normal rules for
```

Comment:
So just to be clear, given `extension E on T {}` and `E x = ...` then `x is T` will promote to `T`, since `T <: E`?

Reply:
Yes.